skorch/callbacks/base.py

```diff
@@ -28,32 +28,26 @@ def initialize(self):
         """
         return self
 
-    def on_train_begin(self, net,
-                       X=None, y=None, **kwargs):
+    def on_train_begin(self, net, X=None, y=None, **kwargs):
         """Called at the beginning of training."""
 
-    def on_train_end(self, net,
-                     X=None, y=None, **kwargs):
+    def on_train_end(self, net, X=None, y=None, **kwargs):
         """Called at the end of training."""
 
-    def on_epoch_begin(self, net,
-                       dataset_train=None, dataset_valid=None, **kwargs):
+    def on_epoch_begin(self, net, dataset_train=None, dataset_valid=None, **kwargs):
         """Called at the beginning of each epoch."""
 
-    def on_epoch_end(self, net,
-                     dataset_train=None, dataset_valid=None, **kwargs):
+    def on_epoch_end(self, net, dataset_train=None, dataset_valid=None, **kwargs):
         """Called at the end of each epoch."""
 
-    def on_batch_begin(self, net,
-                       X=None, y=None, training=None, **kwargs):
+    def on_batch_begin(self, net, batch=None, training=None, **kwargs):
```

Comment:
Agreed. Not every user might follow the changelog closely, we could wrap the `notify` call with an exception handler and in case of `on_batch_begin` we could raise the original exception in addition to a note informing the user about the change in signature and how to fix it. This could stay there for one release.

Reply:
I took a stab at this but it turns out not to be so simple. The reason is that we always have `**kwargs` in the signature, so even if someone writes a wrong method like `def on_batch_begin(net, X=None, y=None, **kwargs)`, it would not immediately lead to an error. The error occurs only when they do _something_ with X or y, because these will be None instead of the expected values. But this _something_ can be anything.

In my attempt, I just caught any kind of `Exception` and raised a `TypeError` from it that pointed at the possible error source. However, this broke other tests, which checked for specific errors (for instance, using scoring with a non existing function name). We don't want that to happen.

The problem is thus that we don't know what error will occur but we also don't want to catch all kinds of errors.

We could theoretically inspect the arguments inside the `notify`-wrapper. But there I would be afraid that the performance penalty could be too big. Furthermore, we want to allow the users to be very liberal in how they call their `on_*` methods (e.g. only passing the arguments that they need, which might not even include `batch`), so it's not possible to do a strict check.

This leaves me with no idea how to implement the suggestion. Do you have any ideas?

Reply:
OK this is fair, maybe we can issue a warning on the first call of `notify('on_batch_begin')` for callbacks that aren't ours that indicates that things might have changed? We can deprecate this warning in the next or a following release and you could always filter the warning if you are getting annoyed by it.

Reply:
So basically on the very first call to `notify`, you'd like to go through all callbacks and check whether they are an instance of any built-in callback, and if at least one of them isn't, issue a warning? I'm not sure if this is not overkill, given that the warning will mostly be a false alarm.

Reply:
Oh, no, more like `if not cb.__module__.startswith('skorch'): issueWarning()`.